pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -1208,6 +1231,25 @@ func (og *operationGenerator) GenerateUnmapVolumeFunc(
 		// plugins/kubernetes.io/{PluginName}/volumeDevices/{volumePluginDependentPath}/{podUID}
 		globalUnmapPath := volumeToUnmount.DeviceMountPath
 
+		// Mark the device as uncertain to make sure kubelet calls UnmapDevice again in all the "return err"
+		// cases below. The volume is marked as fully un-mapped at the end of this function, when everything
+		// succeeds.
+		markVolumeOpts := MarkVolumeOpts{
+			PodName:             volumeToUnmount.PodName,
+			PodUID:              volumeToUnmount.PodUID,
+			VolumeName:          volumeToUnmount.VolumeName,
+			OuterVolumeSpecName: volumeToUnmount.OuterVolumeSpecName,
+			VolumeGidVolume:     volumeToUnmount.VolumeGidValue,
+			VolumeSpec:          volumeToUnmount.VolumeSpec,
+			VolumeMountState:    VolumeMountUncertain,
+		}
+		markVolumeUncertainErr := actualStateOfWorld.MarkVolumeMountAsUncertain(markVolumeOpts)
+		if markVolumeUncertainErr != nil {
```

Comment:
I think this could mean - if a pod running on a node has one or more of volume failing to unmount, then pod can still be deleted because volume is being marked as `uncertain` and hence `GetMountedVolumesForPod` won't find uncertain volumes?

Reply:
good point!

In current logic, pod will not be deleted though. GetMountedVolumesForPod will not find uncertain volumes, but it also check volumepaths from disk directly by calling getMountedVolumePathListFromDisk.

Now thinking this, for pod deletion, I am thinking a safer approach for GetMountedVolumesForPod should include uncertain state..?

Reply:
> n current logic, pod will not be deleted though. GetMountedVolumesForPod will not find uncertain volumes, but it also check volumepaths from disk directly by calling getMountedVolumePathListFromDisk.

yeah I don't trust `getMountedVolumePathListFromDisk` completely. Also, what happens if driver created no mount points but `unpublish` still failed?

Reply:
Adding uncertain volumes to `GetMountedVolumesForPod` will run stats on them and could start pods. This needs to be smarter, perhaps extending only `podVolumesExist` to uncertain volumes.

Reply:
I am trying to remember - in which cases we don't want an uncertain volume to be added in  `GetMountedVolumesForPod`? This would potentially prevent a pod from being deleted in some cases - but may be that is a good thing?

Reply:
When kubelet waits for all volumes to be mounted, it calls `verifyVolumesMountedFunc`, which calls `volumemanager.getUnmountedVolumes()`, which then calls  `GetMountedVolumesForPod`. If uncertain volumes are returned from `GetMountedVolumesForPod`, kubelet will interpret it as the volumes are mounted and it will start the pod.

Reply:
we could have separate logic in these cases.
1. During pod start, when checking whether all volumes are already mounted, this check should not include uncertain volumes
2. During pod deletion, when checking whether volumes are unmounted, it should include uncertain volumes.

Reply:
Yes I think that may work. I am thinking we should leave `GetMountedVolumesForPod` alone because it is also called from other places but may be introduce another function which would include uncertain and mounted volumes before allowing pod deletion.

Reply:
/lgtm cancel
This needs to be reworked as suggested above.

Reply:
I added `GetPossiblyMountedVolumesForPod` for the case when `podVolumesExist` needs to know when all volumes were really unmounted (and are not uncertain).
@gnufied @jingxu97, PTAL